### PR TITLE
fix #125

### LINKED ISSUE #125: phantom resullt details
For certain violations, the report includes completely unrelated result details.

[repro](https://s.zazuko.com/3gCZpze):

```turtle
# shapes
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix ex: <https://example.org/> .

ex:personShape a sh:NodeShape ;
	sh:targetClass ex:Person ;
	sh:property [
		sh:path ex:address ;
		sh:message "ex:city should be sh:IRI" ;
		sh:node [
			sh:path ex:city ;
                        sh:nodeKind sh:IRI
		] ;
	] .

ex:sizeShape 
	sh:targetObjectsOf ex:size ;
	sh:or (
                [ sh:hasValue ex:small ]
		[ sh:hasValue ex:big ]
	) .
```

```turtle
# data
@prefix ex: <https://example.org/> .

ex:person1 a ex:Person ; ex:address ex:address1 .

ex:address1 ex:city "London" .

ex:tv ex:size ex:big .
```

A phantom violation:
```turtle 
    sh:focusNode <https://example.org/big> ;
    sh:resultMessage "Value must be <https://example.org/small>" ;
```
appears as a detail of the correct main violation ("ex:city should be sh:IRI" ).

Notice that the phantom result disappears if we swap `small` and `big` in the shape. Likely it is a temporary result of the `sh:or` evaluation, which should be discarded as soon as the second alternative satisfies the constraint.

Also worth noting that the phantom result is not there if we fix the main issue  ("ex:city should be sh:IRI" ).

### PR DESCRIPTION
**Pull request by giacomociti**
still a draft, looking for early feedback and advice on additional testing

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -67,16 +67,20 @@ class SHACLValidator {
   }
 
   // Exposed to be available from validation functions as `SHACL.nodeConformsToShape`
-  nodeConformsToShape(focusNode, shapeNode) {
+  nodeConformsToShape(focusNode, shapeNode, engine = this.validationEngine.clone()) {
```
Comment (tpluscode) [inline index.js:70]: Why clone?

**Review by tpluscode — CHANGES_REQUESTED** [review]
For testing, check the directory `test/data/data-shapes/custom`. I would place additional test cases there.

File: index.js (hunk 1 above)
Comment (giacomociti) [reply index.js:70]: using a fresh instance of validation engine we avoid side effects (collection of result details) on the current instance.
Only when calling from a `sh:node` constraint we do want to affect the current instance, in the other cases (`sh:and`, `sh:or`, `sh:not`...) we are only interested in the boolean result and we don't care about the details.

Comment (tpluscode) [reply index.js:70]: Hm, I wonder if that is entirely true. I was always thinking that to improve the error messages in complex scenarios would require a smarter analysis of the nested results. But maybe it's better to reorganise the shapes and write good, targeted messages...

**Comment by giacomociti** [comment]
> For testing, check the directory `test/data/data-shapes/custom`. I would place additional test cases there.

I'm afraid we cannot rely on the machinery for official tests: the spec is agnostic wrt `sh:detail`. According to the [test suite documentation](https://w3c.github.io/data-shapes/data-shapes-test-suite/#Validate):

>  Before comparing the results graph with the expected results, the actual validation report must not include any "nested" results, via sh:details. 

and in fact details are [removed ](https://github.com/zazuko/rdf-validate-shacl/blob/3bf052ca1ab54f06d99d403d96bf0c30e7913e5f/test/data-shapes_tests.js#L137) before comparing the actual with the expected results

File: index.js (hunk 1 above)
Comment (giacomociti) [reply index.js:70]: admittedly, this PR is a bit of hacking trying to fix the anomaly. Redesigning all the result details to provide more insights for violations of logical constraints is intriguing but beyond the scope of this PR.

**Review by tpluscode — APPROVED** [review]

**Pull request merged**